### PR TITLE
notes

### PR DESCRIPTION
**Pull request by invaliduser**
adds `note_created`, `note_updated`, `notes_viewed` events

### CODE REVIEW

File: src/transformer/events/core/note_created.php [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+ */
+
+function note_created(array $config, \stdClass $event) {
+  global $CFG;
+  $repo = $config['repo'];
+  if (isset($event->objecttable) && isset($event->objectid)) {
+    $event_object = $repo->read_record_by_id($event->objecttable, $event->objectid);
+  } else {
+    $event_object = array();
+  }
```
Comment (milt) [inline src/transformer/events/core/note_created.php]: it's a little weird to keep the generic debug code in here, if you look at our other events we generally name things a little better so they are comprehensible later on. the event object is a note, call it that. The whole if construct is superfluous if you know what event you are working with.

File: src/transformer/events/core/note_created.php [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+
+  //all three here may not exist
+  $actor=$repo->read_record_by_id('user',$event->userid);
+  $subject=$repo->read_record_by_id('user',$event->relateduserid); 
+  $course = (isset($event->courseid) && $event->courseid != 0)
+    ? $repo->read_record_by_id('course', $event->courseid)
+    : null;
+  $lang = utils\get_course_lang(($course
+                                 ? $course
+                                 : $repo->read_record_by_id("course",1)));
```
Comment (milt) [inline src/transformer/events/core/note_created.php]: same stuff about lang as before, use course and fall back to source

File: src/transformer/events/core/note_created.php [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+
+  $statement = [
+    'actor' => utils\get_user($config,$actor),
+    'verb' => ['id' => 'http://activitystrea.ms/create',
+               'display' => ['en' => 'Created']
```
Comment (milt) [inline src/transformer/events/core/note_created.php]: why are you hardcoding lang

File: src/transformer/events/core/note_updated.php [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+ */
+
+function note_updated(array $config, \stdClass $event) {
+    global $CFG;
+    $repo = $config['repo'];
+    if (isset($event->objecttable) && isset($event->objectid)) {
+        $event_object = $repo->read_record_by_id($event->objecttable, $event->objectid);
+    } else {
+        $event_object = array();
+    }
```
Comment (milt) [inline src/transformer/events/core/note_updated.php]: see https://github.com/yetanalytics/moodle-logstore_xapi/pull/52/files#r1836969306

File: src/transformer/events/core/note_updated.php [hunk 5]
```diff
@@ -0,0 +1,89 @@
...
+        ? $repo->read_record_by_id('course', $event->courseid)
+        : null;
+    $lang = utils\get_course_lang(($course
+                                   ? $course
+                                   : $repo->read_record_by_id("course",1)));
+
+    $statement = [
+        'actor' => utils\get_user($config,$user),
+        'verb' => ['id' => 'http://activitystrea.ms/update',
+                   'display' => ['en' => 'Updated']
```
Comment (milt) [inline src/transformer/events/core/note_updated.php]: hardcoded lang

File: src/transformer/events/core/notes_viewed.php [hunk 6]
```diff
@@ -0,0 +1,82 @@
...
+        ? $repo->read_record_by_id('course', $event->courseid)
+        : null;
+    $lang = utils\get_course_lang(($course
+                                   ? $course
+                                   : $repo->read_record_by_id("course",1)));
+
+    $statement = [
+        'actor' => utils\get_user($config,$user),
+        'verb' => ['id' => 'http://id.tincanapi.com/verb/viewed',
+                   'display' => ['en' => 'Viewed']
```
Comment (milt) [inline src/transformer/events/core/notes_viewed.php]: hardcoded lang

File: tests/core/note_created/user_created_note/data.json [hunk 7]
```diff
@@ -0,0 +1,31 @@
+{
+  "course": [
+    {
+      "fullname": "test_course_name",
+      "id": 1,
+      "lang": "en",
+      "summary": "test_course_summary"
+    }
+  ],
```
Comment (milt) [inline tests/core/note_created/user_created_note/data.json]: You don't need to put the course in here.

File: src/transformer/events/core/note_created.php [hunk 8]
```diff
@@ -0,0 +1,89 @@
...
+        'name' => [$lang => $event_object->subject],
+        'description' => [$lang => $event_object->content],
+        'type' =>  'http://activitystrea.ms/note',
+        'extensions' => [
+          "https://xapi.edlm/profiles/edlm-lms/concepts/activity-extensions/note-type"=> "course",
+          "https://xapi.edlm/profiles/edlm-lms/concepts/activity-extensions/note-subject" =>
+            utils\get_user($config,$subject)
+        ]
+      ],
+    ],
```
Comment (milt) [inline src/transformer/events/core/note_created.php]: this should be a function you call from the various note events

File: src/transformer/events/core/note_created.php [hunk 9]
```diff
@@ -0,0 +1,89 @@
...
+  $statement = [
+    'actor' => utils\get_user($config,$actor),
+    'verb' => ['id' => 'http://activitystrea.ms/create',
+               'display' => ['en' => 'Created']
+    ],
+    'object' => [
+      'id' => $config['app_url'].'/notes/view.php?id='.$event->id,
+      'definition' => [
+        'name' => [$lang => $event_object->subject],
+        'description' => [$lang => $event_object->content],
```
Comment (milt) [inline src/transformer/events/core/note_created.php]: As with some other things, we should manually check to see if these have HTML tags in them in production that need to be stripped

File: tests/core/note_created/user_created_note/user_created_note_test.php [hunk 10]
```diff
@@ -0,0 +1,70 @@
...
+     * @return string
+     */
+    protected function get_plugin_name() {
+        return "core";
+    }
+
+    /**
+     * Appease auto-detecting of test cases. xapi_test_case has default test cases.
+     *
+     * @covers ::attempt_submitted
```
Comment (milt) [inline tests/core/note_created/user_created_note/user_created_note_test.php]: we don't really know what these do, but this one is wrong

File: tests/core/note_updated/user_updated_note/data.json [hunk 11]
```diff
@@ -0,0 +1,31 @@
+{
+  "course": [
+    {
+      "fullname": "test_course_name",
+      "id": 1,
+      "lang": "en",
+      "summary": "test_course_summary"
+    }
+  ],
```
Comment (milt) [inline tests/core/note_updated/user_updated_note/data.json]: not needed

File: tests/core/note_updated/user_updated_note/user_updated_note_test.php [hunk 12]
```diff
@@ -0,0 +1,70 @@
...
+     * @return string
+     */
+    protected function get_plugin_name() {
+        return "core";
+    }
+
+    /**
+     * Appease auto-detecting of test cases. xapi_test_case has default test cases.
+     *
+     * @covers ::attempt_submitted
```
Comment (milt) [inline tests/core/note_updated/user_updated_note/user_updated_note_test.php]: annotation is wrong

File: tests/core/notes_viewed/user_viewed_notes/data.json [hunk 13]
```diff
@@ -0,0 +1,24 @@
+{
+  "course": [
+    {
+      "fullname": "test_course_name",
+      "id": 1,
+      "lang": "en",
+      "summary": "test_course_summary"
+    }
+  ],
```
Comment (milt) [inline tests/core/notes_viewed/user_viewed_notes/data.json]: don't need course

File: tests/core/notes_viewed/user_viewed_notes/user_viewed_notes_test.php [hunk 14]
```diff
@@ -0,0 +1,70 @@
...
+     * @return string
+     */
+    protected function get_plugin_name() {
+        return "core";
+    }
+
+    /**
+     * Appease auto-detecting of test cases. xapi_test_case has default test cases.
+     *
+     * @covers ::attempt_submitted
```
Comment (milt) [inline tests/core/notes_viewed/user_viewed_notes/user_viewed_notes_test.php]: wrongo

Comment (milt) [inline tests/core/notes_viewed/user_viewed_notes/user_viewed_notes_test.php]: wrongo

**Review by milt — CHANGES_REQUESTED** [review]
See various notes

**Review by cliffcaseyyet — APPROVED** [review]

**Review by milt — APPROVED** [review]

**Pull request merged**